docs/Toy/toy/interpreter.py

```diff
@@ -0,0 +1,150 @@
+from io import StringIO
+import operator
+from typing import Any
+from itertools import accumulate
+
+from dataclasses import dataclass
+
+from xdsl.dialects.builtin import TensorType, VectorType, ModuleOp
+from xdsl.interpreter import Intepreter, IntepretationError, Functions
+
+from . import dialect as toy
+
+
+def run_toy_func(intepreter: Intepreter, name: str,
+                 args: tuple[Any, ...]) -> tuple[Any, ...]:
+    for op in intepreter.module.regions[0].blocks[0].ops:
+        if isinstance(op, toy.FuncOp) and op.sym_name.data == name:
+            return run_func(intepreter, op, args)
+
+    raise IntepretationError(f'Could not find toy function with name: {name}')
+
+
+toy_ft = Functions()
+
+
+@dataclass
+class Tensor:
+    data: list[float]
+    shape: list[int]
+
+    def __format__(self, __format_spec: str) -> str:
+        prod_shapes: list[int] = list(
+            accumulate(reversed(self.shape), operator.mul))
+        assert prod_shapes[-1] == len(self.data)
+        result = '[' * len(self.shape)
+
+        for i, d in enumerate(self.data):
+            if i:
+                n = sum(not i % p for p in prod_shapes)
+                result += ']' * n
+                result += ', '
+                result += '[' * n
+            result += f'{d}'
+
+        result += ']' * len(self.shape)
+        return result
+
+
+@toy_ft.register(toy.PrintOp)
```

Comment:
This is the main thing I'm not sure about, and would love feedback for. I have some alternatives in mind, we could go for a similar annotation to `RewritePattern`, meaning on an instance of `Functions` (name also very much up for discussion), or similar to `irdl` annotations on the class, that register all the functions of the correct type.

I like the API in this PR because it forces statelessness of the functions object itself, we want all state to go through the environment of the interpreter.

Reply:
Hmm I think this is quite a nice design actually. Personally, I would have had this as part of the interpreter, but separating them is actually a good idea, IMO.